asdf/config.py

```diff
@@ -0,0 +1,237 @@
+        every provider registered as an entry point.
+
+        Returns
+        -------
+        list of asdf.AsdfConverterProvider
+        """
+        if self._converter_providers is None:
+            with self._lock:
+                if self._converter_providers is None:
+                    self._converter_providers = get_registered_converter_providers()
+
+        return self._converter_providers
+
+    @property
+    def converters(self):
+        """
+        Get the list of `AsdfConverter` instances provided
+        by the configured converter_providers.
+
+        Returns
+        -------
+        list of asdf.AsdfConverter
+        """
+        if self._converters is None:
+            with self._lock:
+                if self._converters is None:
+                    self._converters = create_converters(self.converter_providers)
+
+        return self._converters
+
+    @property
+    def converter_collection(self):
+        """
+        Get the `ConverterCollection` for the configured converters.
+
+        Returns
+        -------
+        asdf.converter.ConverterCollection
+        """
+
+        if self._converter_collection is None:
+            with self._lock:
+                if self._converter_collection is None:
+                    self._converter_collection = ConverterCollection(self.converters)
+
+        return self._converter_collection
+
+    @property
+    def validate_on_read(self):
+        """
+        Get configuration that controls schema validation of
+        ASDF files on read.
+
+        Returns
+        -------
+        bool
+        """
+        return self._validate_on_read
+
+    def _replace(
+        self,
+        converter_providers=NotSet,
+        validate_on_read=NotSet,
+    ):
+        if converter_providers is NotSet:
+            converter_providers = self._converter_providers
+            converters = self._converters
+            converter_collection = self._converter_collection
+        else:
+            converters = None
+            converter_collection = None
+
+        if validate_on_read is NotSet:
+            validate_on_read = self._validate_on_read
+
+        return AsdfConfig(
+            converter_providers=converter_providers,
+            converters=converters,
+            converter_collection=converter_collection,
+            validate_on_read=validate_on_read,
+        )
+
+
+class _ConfigLocal(threading.local):
+    def __init__(self):
+        self.config_stack = []
+
+
+_global_config = AsdfConfig()
+_local = _ConfigLocal()
+
+
+def get_config():
+    """
+    Get the current config, which may have been altered by
+    one or more containing `AsdfConfigContext`.
+
+    Returns
+    -------
+    AsdfConfig
+    """
+    if len(_local.config_stack) == 0:
+        return _global_config
+    else:
+        return _local.config_stack[-1]
+
+
+def configure(
+    converter_providers=NotSet,
+    validate_on_read=NotSet,
+):
```

Comment:
`None` does in fact have a distinct meaning here -- it means "load the providers registered with entry points".  The purpose of `NotSet` is to allow users to configure options piecemeal.  If the defaults were `None`, then this call:

```python
asdf.configure(validate_on_read=True)
```

would cause the current value of converter_providers to be wiped out and reloaded from entry points, which I doubt is what the user wants.

An alternative way to implement this same interface is to use `**kwargs`:
 
```python
def configure(**kwargs):
  if "converter_providers" in kwargs:
    converter_providers = kwargs["converter_providers"]
  else:
    converter_providers = <the previous value>
```

but I find this significantly less usable because IDEs won't be able to detect and list possible keywords, the user will have to dig into the docstring.

Reply:
This may not be clear from the documentation, but the intended behavior of this method is to "make a selective update to the current configuration", not "set all configuration options and use defaults for anything not explicitly set".

Reply:
I don't think that behaves the way we'd need, for example this:
```python
In [24]: configure(converter_providers=None)
cp default
vor default
```
gives the same result as this:
```python
In [25]: configure()
cp default
vor default
```
What I'm looking for is the following behavior:
```python
In [26]: configure()
cp use previous value
vor use previous value

In [26]: configure(converter_providers=[CustomConverterProvider()])
cp set to [CustomConverterProvider instance]
vor use previous value

In [26]: configure(converter_providers=None)
cp reload providers from entry points
vor use previous value

In [26]: configure(validate_on_read=False)
cp use previous value
vor set to False
```

Reply:
Yes, that's right, at least in the case of `converter_providers`.  Does it makes sense to make that a consistent response to `None` across all options?  `validate_on_read` isn't currently coded up that way, but it could be.

Reply:
Yes, it probably would be good to be consistent with that.

Reply:
Ok, this is clearer for me now.